src/language/variables.md

```diff
@@ -38,6 +38,34 @@ String name = 'Bob';
   of using `var`, rather than type annotations, for local variables.
 {{site.alert.end}}
 
+## Null safety
+
+The Dart language enforces sound null safety.
+
+Null safety prevents errors that result from unintentional access
+of variables set to `null`.
+
+For example, if a method expects an integer but receives `null`,
+your app causes a runtime error.
+This type of error, a null dereference error, can be difficult to debug.
+
+With sound null safety, all variables require a value.
+This means Dart considers all variables _non-nullable_.
```

Comment:
This isn't true. This is fine:

```dart
var x = null;
```

Dart considers `null` to be a totally valid value and a perfectly fine value to store in variables. It's also perfectly happy to implicitly initialize variables to `null` on your behalf if you don't initialize it:

```dart
String? maybeString;
print(maybeString == null); // Prints "true".
```

The invariant Dart gives you is:

1. It's never possible to observe an uninitialized variable. It may get default initialized to `null`, but you'll never ever see garbage bits.
2. If a variable doesn't allow `null`, the language forces you to initialize it to some explicit value, since there is no default initialization for non-nullable types (unlike, say, Go or C#, where all types support default initialization).
3. If an expression or variable's type is non-nullable, then it will never evaluate to `null`.
4. If an expression or variable can be `null`, you can't access any properties or call any methods on it (unless those methods are valid for `null`, like `toString()`).

I'm not sure how best to summarize that here.

Reply:
I think I have a summary:
```markdown
In Dart, variables must be initialized before they can be used.
If a variable is not initialized, Dart assigns it the default value of null.
However, it is never possible to observe an uninitialized variable.
Dart also prevents you from accessing properties or calling methods
on a variable that equals `null`.
Unless you design the method to handle `null` values, you can't use that method
on a variable that equals `null`.
```

Reply:
> If a variable is not initialized, Dart assigns it the default value of null.

That will throw off some readers who will think it means that that variables aren't sound with respect to `null`. I'd summarize it as:

* In Dart, the compiler requires you to initialize a variable before it will let you use it.
* If the variable's type is nullable, the language will default initialize it to `null` for you. Otherwise, you have to explicitly initialize it yourself before you can use it.
* You can't call a method or getter on an expression whose type is nullable (unless it's one of a couple of methods that are defined on `null` too).

You can't really "design the method to handle `null`". There are a couple members defined in the Dart core libraries on the `Null` class that we add but end users can't add anything else to that class. (You *can* define extensions on nullable types, but that's probably out of scope here and not really related to null safety.)